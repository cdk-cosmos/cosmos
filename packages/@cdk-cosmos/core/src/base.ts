import { IConstruct, CfnResource } from '@aws-cdk/core';
import { Construct, ConstructNode } from '@aws-cdk/core/lib/construct-compat';
import { Stack, StackProps } from '@aws-cdk/core/lib/stack';

const COSMOS_PARTITION = 'COSMOS_PARTITION';
const COSMOS_VERSION = 'COSMOS_VERSION';
const DEFAULT_PATTERN = '{Resource}+';
const STACK_PATTERN = '{Partition}{Cosmos}{Galaxy}?{SolarSystem}?{Version}?{Type}';

declare module '@aws-cdk/core/lib/construct-compat' {
  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface ConstructNode {
    id: string;
    type: string;
    pattern: string;
  }
}

ConstructNode.prototype.type = 'Resource';

export interface BaseStackProps extends StackProps {
  partition?: string;
  version?: string;
  type: string;
  disableCosmosNaming?: boolean;
}

export class BaseStack extends Stack {
  private disableCosmosNaming: boolean;

  constructor(scope: Construct, id: string, props: BaseStackProps) {
    const { partition, version, type, disableCosmosNaming = false } = props;

    const stackName = generateLogicalId({
      scopes: [...getScopes(scope), { key: type, value: id }, { key: 'Type', value: type }],
      pattern: STACK_PATTERN,
      partition: partition,
      version: version,
    });

    super(scope, id, {
      stackName,
      ...props,
    });

    this.node.type = type;
    if (partition) this.node.setContext(COSMOS_PARTITION, partition);
    if (version) this.node.setContext(COSMOS_VERSION, version);
    this.disableCosmosNaming = disableCosmosNaming;
  }

  public allocateLogicalId(scope: CfnResource): string {
    if (this.disableCosmosNaming) return super.allocateLogicalId(scope);

    return generateLogicalId({
      scopes: getScopes(scope),
      pattern: getPattern(scope) || DEFAULT_PATTERN,
      partition: scope.node.tryGetContext(COSMOS_PARTITION),
      version: scope.node.tryGetContext(COSMOS_VERSION),
    });
  }
}

interface IScope {
  key: string;
  value: string;
}

interface GenerateLogicalIdProps {
  scopes: IScope[];
  pattern: string;
  partition?: string;
  version?: string;
}

export function generateLogicalId(props: GenerateLogicalIdProps): string {
  const { scopes, pattern, partition, version } = props;

  if (partition) scopes.push({ key: 'Partition', value: partition });
  if (version) scopes.push({ key: 'Version', value: version });

  const selectedIds = selectScoped(scopes, pattern)
    .map(x => x.value)
    .reduce(removeDupes, [])
    .filter(removeHidden)
    .map(removeNonAlphanumeric);

  return selectedIds.join('').slice(0, 240);
}

function selectScoped(scopes: IScope[], pattern: string): IScope[] {
  const regex = /{(\w+)}([*+?])?/gm;
  const results: IScope[] = [];
  let segment: RegExpExecArray | null;
  while ((segment = regex.exec(pattern)) !== null) {
    const key: string = segment[1];
    const selector: string | undefined = segment[2];

    switch (selector) {
      case '*': {
        const selected = scopes.filter(x => x.key === key);
        if (selected.length) results.push(...selected);
        break;
      }
      case '+': {
        if (!scopes.some(x => x.key === key)) throw new Error(`No ${key} Scope found`);
        results.push(...scopes.filter(x => x.key === key));
        break;
      }
      case '?': {
        const selected = scopes.filter(x => x.key === key).pop();
        if (selected) results.push(selected);
        break;
      }
      default: {
        const selected = scopes.filter(x => x.key === key).pop();
        if (!selected) throw new Error(`No ${key} Scope found`);
        results.push(selected);
      }
    }
  }

  return results;
}

function getScopes(scope: IConstruct): IScope[] {
  const scopes = scope.node.scopes
    .filter(x => x.node.id && x.node.type)
    .map(x => ({
      key: x.node.type,
      value: x.node.id,
    }));

  if (scope.node.type) scopes.push({ key: 'Type', value: scope.node.type });

  return scopes;
}

function getPattern(scope: IConstruct): string | undefined {
  return scope.node.scopes
    .map(x => x.node.pattern)
    .filter(x => x)
    .pop();
}

function removeDupes(result: string[], component: string): string[] {
  if (result.length === 0 || !result[result.length - 1].endsWith(component)) {
    result.push(component);
  }

  return result;
}

function removeHidden(value: string): boolean {
  if (value === 'Default' || value === 'Resource') return false;
  return true;
}

function removeNonAlphanumeric(s: string): string {
  return s.replace(/[^A-Za-z0-9]/g, '');
}

// ---- ---- ---- ----

// /**
//  * Resources with this ID are hidden from humans
//  *
//  * They do not appear in the human-readable part of the logical ID,
//  * but they are included in the hash calculation.
//  */
// const HIDDEN_FROM_HUMAN_ID = 'Resource';

// /**
//  * Resources with this ID are complete hidden from the logical ID calculation.
//  */
// const HIDDEN_ID = 'Default';

// const PATH_SEP = '/';

// const HASH_LEN = 8;
// const MAX_HUMAN_LEN = 240; // max ID len is 255
// const MAX_ID_LEN = 255;

// /**
//  * Calculates a unique ID for a set of textual components.
//  *
//  * This is done by calculating a hash on the full path and using it as a suffix
//  * of a length-limited "human" rendition of the path components.
//  *
//  * @param components The path components
//  * @returns a unique alpha-numeric identifier with a maximum length of 255
//  */
// export function makeUniqueId(components: string[]) {
//   components = components.filter(x => x !== HIDDEN_ID);

//   if (components.length === 0) {
//     throw new Error('Unable to calculate a unique id for an empty set of components');
//   }

//   // Lazy require in order to break a module dependency cycle
//   const unresolvedTokens = components.filter(c => unresolved(c));
//   if (unresolvedTokens.length > 0) {
//     throw new Error(`ID components may not include unresolved tokens: ${unresolvedTokens.join(',')}`);
//   }

//   // top-level resources will simply use the `name` as-is in order to support
//   // transparent migration of cloudformation templates to the CDK without the
//   // need to rename all resources.
//   if (components.length === 1) {
//     // we filter out non-alpha characters but that is actually a bad idea
//     // because it could create conflicts ("A-B" and "AB" will render the same
//     // logical ID). sadly, changing it in the 1.x version line is impossible
//     // because it will be a breaking change. we should consider for v2.0.
//     // https://github.com/aws/aws-cdk/issues/6421
//     const candidate = removeNonAlphanumeric(components[0]);

//     // if our candidate is short enough, use it as is. otherwise, fall back to
//     // the normal mode.
//     if (candidate.length <= MAX_ID_LEN) {
//       return candidate;
//     }
//   }

//   const hash = pathHash(components);
//   const human = removeDupes(components)
//     .filter(x => x !== HIDDEN_FROM_HUMAN_ID)
//     .map(removeNonAlphanumeric)
//     .join('')
//     .slice(0, MAX_HUMAN_LEN);

//   return human + hash;
// }

// /**
//  * Take a hash of the given path.
//  *
//  * The hash is limited in size.
//  */
// function pathHash(path: string[]): string {
//   const md5 = crypto
//     .createHash('md5')
//     .update(path.join(PATH_SEP))
//     .digest('hex');
//   return md5.slice(0, HASH_LEN).toUpperCase();
// }

// /**
//  * Removes all non-alphanumeric characters in a string.
//  */
// function removeNonAlphanumeric(s: string) {
//   return s.replace(/[^A-Za-z0-9]/g, '');
// }

/**
 * Remove duplicate "terms" from the path list
 *
 * If the previous path component name ends with this component name, skip the
 * current component.
 */
// function removeDupes(path: string[]): string[] {
//   const ret = new Array<string>();

//   for (const component of path) {
//     if (ret.length === 0 || !ret[ret.length - 1].endsWith(component)) {
//       ret.push(component);
//     }
//   }

//   return ret;
// }
