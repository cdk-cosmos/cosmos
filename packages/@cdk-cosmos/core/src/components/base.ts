import { CfnResource } from '@aws-cdk/core';
import { Construct } from '@aws-cdk/core/lib/construct-compat';
import { NetworkBuilder } from '@aws-cdk/aws-ec2/lib/network-util';
import { Stack, StackProps } from '@aws-cdk/core/lib/stack';
import { PATTERN, COSMOS_PARTITION, COSMOS_VERSION, COSMOS_NETWORK_BUILDER, AWS_ENV } from '../helpers/constants';
import { generateNodeId, generateScopeId, generateSingletonId } from '../helpers/generate-scope-id';

export interface BaseStackOptions extends StackProps {
  partition?: string;
  version?: string;
  disableCosmosNaming?: boolean;
  cidr?: string;
}

export interface BaseStackProps extends BaseStackOptions {
  type: string;
}

export class BaseStack extends Stack {
  private readonly disableCosmosNaming: boolean;
  public readonly networkBuilder?: NetworkBuilder;

  constructor(scope: Construct, id: string, props: BaseStackProps) {
    const { partition, version, type, disableCosmosNaming = false, env, cidr } = props;
    const stackName = generateNodeId({ scope, id, type, partition, version, pattern: PATTERN.STACK });

    super(scope, id, {
      stackName,
      env: scope.node.tryGetContext(AWS_ENV),
      ...props,
    });

    this.node.type = type;
    this.disableCosmosNaming = disableCosmosNaming;

    if (partition) this.node.setContext(COSMOS_PARTITION, partition);
    if (version) this.node.setContext(COSMOS_VERSION, version);
    if (env) this.node.setContext(AWS_ENV, env);

    if (cidr) this.networkBuilder = new NetworkBuilder(cidr);
    if (this.networkBuilder) this.node.setContext(COSMOS_NETWORK_BUILDER, this.networkBuilder);
    else this.networkBuilder = this.node.tryGetContext(COSMOS_NETWORK_BUILDER);
  }

  public allocateLogicalId(scope: CfnResource): string {
    if (this.disableCosmosNaming) return super.allocateLogicalId(scope);
    const id = generateScopeId({ scope, defaultPattern: PATTERN.RESOURCE });
    return removeNonAlphanumeric(id);
  }
}

declare module '@aws-cdk/core/lib/construct-compat' {
  interface ConstructNode {
    id: string;
    type?: string;
    pattern?: string;
  }

  interface Construct {
    nodeId(id?: string, delimiter?: string, pattern?: string, type?: string): string;
    singletonId(id?: string, delimiter?: string, type?: string): string;
  }
}

Construct.prototype.nodeId = function(id, delimiter, pattern, type): string {
  return generateNodeId({ scope: this, pattern, id, type, delimiter });
};

Construct.prototype.singletonId = function(id, delimiter, type): string {
  return generateSingletonId({ scope: this, id, type, delimiter });
};

const removeNonAlphanumeric = (s: string): string => {
  return s.replace(/[^A-Za-z0-9]/g, '');
};
