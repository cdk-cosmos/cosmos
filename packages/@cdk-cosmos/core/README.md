# Cosmos Core Package

This Library contains the base components for the CDK Cosmos. It defiles the Core Classes and the Extension Classes for the Core.
All other @cdk-cosmos packages will build on top of this core concept.

## Basic Understanding

Cosmos is broken down into 3 scopes (Bubbles)

- Cosmos: Top level scope of the entire infrastructure, mostly singleton shared resources.
- Galaxy: Account level scope that defines infrastructure for an aws account.
- SolarSystem: Environment level scope that defines infrastructure for an environment with in an aws account (Dev, Tst, Prd).

## Cosmos Core

The Cosmos core contains infrastructure that will be shared by consumers building apps (extensions).

The Core Cosmos is composed of:

- CosmosStack
  - TLD Zone
  - CDK Repo - for the core cdk code
  - CDK Master Role - for builds of the cdk deployments
- GalaxyStack
  - If Cross Account, create Cross Account CDK Role for master to assume
- SolarSystemStack
  - VPC (Created on the Galaxy scope, so that it can be shared by many SolarSystems in the same Galaxy)
  - Zone
- EcsSolarSystemStack
  - Cluster
  - Alb
  - HttpListener
- CiCdSolarSystemStack
  - VPC
  - Zone
  - Cluster
  - Alb
  - HttpListener
  - CDK Pipeline + Code Build for Deployment of the core

### Example

```js
const cosmos = new CosmosStack(app, 'Demo', {
  tld: 'cosmos.com',
  cidr: '10.0.0.0/22',
});

const mgtGalaxy = new GalaxyStack(cosmos, 'Mgt');

const ciCd = new CiCdSolarSystemStack(mgtGalaxy);

const dev = new EcsSolarSystemStack(mgtGalaxy, 'Dev');
```

## Cosmos Extensions

The Extensions are a way of consuming and adding resources in an isolated way. Each Extension has a Portal Property that is used to consume that core infrastructure for that same bubble.

The Extension of the core are:

- CosmosExtensionStack
  - Imports Core Cosmos
  - CDK Repo - for the app cdk code
- GalaxyExtensionStack
  - Imports Core Galaxy
- SolarSystemExtensionStack
  - Imports Core SolarSystem
- EcsSolarSystemExtensionStack
  - Imports Core EcsSolarSystem
- CiCdSolarSystemExtensionStack
  - Imports Core CiCdSolarSystem
  - CDK Pipeline + Code Build for Deployment of the app

### Example

```js
const cosmos = new CosmosExtensionStack(app, 'Demo');

const mgtGalaxy = new GalaxyExtensionStack(cosmos, 'Mgt');

const ciCd = new CiCdSolarSystemExtensionStack(mgtGalaxy);

const dev = new EcsSolarSystemExtensionStack(mgtGalaxy, 'Dev');
```

What is included in the above is just enough for the extension (App) to host its own code and deploy it self. From here we expect consumer to add their own infrastructure to extends from the core, into the respective bubbles.

## Further Use

For more extensive understanding and usage, please look at the docs, aka the [law of the cosmos](https://github.com/cdk-cosmos/law)

Deploy CDKToolKit `npx cdk deploy --app="node_modules/@cosmos-building-blocks/common/lib/cdk-tool-kit-app.js"`