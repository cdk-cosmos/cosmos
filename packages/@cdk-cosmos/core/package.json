{
  "name": "@cdk-cosmos/core",
  "version": "0.2.0",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib/*",
    "package.json",
    "README.md"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/cdk-cosmos/cosmos.git",
    "directory": "packages/@cdk-cosmos/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsc",
    "watch": "tsc -w",
    "test": "jest",
    "lint": "eslint . --ext .js,.ts",
    "format": "prettier --write **/*.ts !**/*.d.ts",
    "clean": "rm -rf lib"
  },
  "dependencies": {
    "@cosmos-building-blocks/common": "^0.2.0"
  },
  "peerDependencies": {
    "@aws-cdk/aws-cloudformation": "^1.23.0",
    "@aws-cdk/aws-codebuild": "^1.23.0",
    "@aws-cdk/aws-codecommit": "^1.23.0",
    "@aws-cdk/aws-codepipeline": "^1.23.0",
    "@aws-cdk/aws-codepipeline-actions": "^1.23.0",
    "@aws-cdk/aws-ec2": "^1.23.0",
    "@aws-cdk/aws-ecr": "^1.23.0",
    "@aws-cdk/aws-ecs": "^1.23.0",
    "@aws-cdk/aws-elasticloadbalancingv2": "^1.23.0",
    "@aws-cdk/aws-elasticloadbalancingv2-targets": "^1.23.0",
    "@aws-cdk/aws-iam": "^1.23.0",
    "@aws-cdk/aws-lambda": "^1.23.0",
    "@aws-cdk/aws-route53": "^1.23.0",
    "@aws-cdk/core": "^1.23.0"
  },
  "gitHead": "173c051038a3acdaa93d6a37d42b6b0546ffd567"
}
