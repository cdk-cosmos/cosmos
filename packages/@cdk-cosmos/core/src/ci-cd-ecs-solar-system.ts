import { Project } from '@aws-cdk/aws-codebuild';
import {
  IEcsSolarSystemCore,
  EcsSolarSystemCoreStack,
  EcsSolarSystemCoreProps,
  ImportedEcsSolarSystemCore,
  EcsSolarSystemExtensionStack,
  IEcsSolarSystemExtension,
} from './ecs-solar-system';
import { IGalaxyExtension, IGalaxyCore } from './galaxy';
import { CdkPipeline } from './components/cdk-pipeline';
import {
  ICiCdSolarSystemCore,
  ICiCdSolarSystemExtension,
  CiCdSolarSystemCoreStackProps,
  CiCdSolarSystemExtensionStackProps,
  CosmosCdkPipeline,
} from './ci-cd-solar-system';

export interface ICiCdEcsSolarSystemCore extends ICiCdSolarSystemCore, IEcsSolarSystemCore {}

export interface CiCdEcsSolarSystemCoreStackProps extends CiCdSolarSystemCoreStackProps, EcsSolarSystemCoreProps {}

export class CiCdEcsSolarSystemCoreStack extends EcsSolarSystemCoreStack implements ICiCdEcsSolarSystemCore {
  readonly deployPipeline: CdkPipeline;
  readonly deployProject: Project;

  constructor(galaxy: IGalaxyCore, props?: CiCdEcsSolarSystemCoreStackProps) {
    super(galaxy, 'CiCd', {
      description: 'Cosmos CiCdEcsSolarSystem: Resources dependant on Ci & Cd, like CodePipelines and CodeDeployments.',
      ...props,
    });

    const { cdkPipelineProps } = props || {};

    this.deployPipeline = new CosmosCdkPipeline(this, 'CdkPipeline', cdkPipelineProps);
    this.deployProject = this.deployPipeline.Deploy;
  }
}

export interface ICiCdEcsSolarSystemExtension extends ICiCdSolarSystemExtension, IEcsSolarSystemExtension {
  portal: ICiCdEcsSolarSystemCore;
}

export interface CiCdEcsSolarSystemExtensionStackProps extends CiCdSolarSystemExtensionStackProps {}

export class CiCdEcsSolarSystemExtensionStack extends EcsSolarSystemExtensionStack
  implements ICiCdEcsSolarSystemExtension {
  readonly portal: ICiCdEcsSolarSystemCore;
  readonly deployPipeline: CdkPipeline;
  readonly deployProject: Project;

  constructor(galaxy: IGalaxyExtension, props?: CiCdEcsSolarSystemExtensionStackProps) {
    super(galaxy, 'CiCd', {
      description:
        'Cosmos CiCdEcsSolarSystem Extension: App resources dependant on Ci & Cd, like CodePipelines and CodeDeployments.',
      ...props,
    });

    const { cdkPipelineProps } = props || {};

    this.deployPipeline = new CosmosCdkPipeline(this, 'CdkPipeline', cdkPipelineProps);
    this.deployProject = this.deployPipeline.Deploy;
  }

  protected getPortal(props?: CiCdSolarSystemExtensionStackProps): ICiCdEcsSolarSystemCore {
    const galaxy = this.node.scope as IGalaxyExtension;
    return new ImportedEcsSolarSystemCore(galaxy.portal, this.node.id, props?.portalProps);
  }
}
