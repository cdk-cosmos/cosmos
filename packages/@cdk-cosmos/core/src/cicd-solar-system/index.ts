export {
  ICiCdSolarSystemCore,
  CiCdSolarSystemCoreStack,
  CiCdEcsSolarSystemCoreStack,
  CiCdSolarSystemCoreStackProps,
} from './cicd-solar-system-core-stack';
export {
  ICiCdSolarSystemExtension,
  CiCdSolarSystemExtensionStack,
  CiCdEcsSolarSystemExtensionStack,
  CiCdSolarSystemExtensionStackProps,
} from './cicd-solar-system-extension-stack';
