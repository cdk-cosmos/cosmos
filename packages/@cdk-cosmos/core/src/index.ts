export {
  Bubble,
  Cosmos,
  Galaxy,
  SolarSystem,
  EcsSolarSystem,
  CiCdSolarSystem,
  Extension,
  CosmosExtension,
  GalaxyExtension,
  SolarSystemExtension,
  EcsSolarSystemExtension,
  CiCdSolarSystemExtension,
} from './interfaces';

export {
  RemoteZone,
  RemoteVpc,
  RemoteCluster,
  RemoteAlb,
  RemoteApplicationListener,
  RemoteCodeRepo,
  RemoteBuildProject,
} from './remote';

export { CosmosStack, CosmosStackProps, ImportedCosmos, CosmosExtensionStack } from './cosmos';

export { GalaxyStack, GalaxyStackProps, ImportedGalaxy, GalaxyExtensionStack } from './galaxy';

export { SolarSystemStack, SolarSystemProps, ImportedSolarSystem, SolarSystemExtensionStack } from './solar-system';

export {
  EcsSolarSystemStack,
  EcsSolarSystemProps,
  ImportedEcsSolarSystem,
  EcsSolarSystemExtensionStack,
} from './ecs-solar-system';

export {
  CiCdSolarSystemStack,
  CiCdStackProps,
  ImportedCiCdSolarSystem,
  CiCdSolarSystemExtensionStack,
} from './ci-cd-solar-system';

export { CdkPipeline, CdkPipelineProps } from './cdk-pipeline';

export { RESOLVE, PATTERN, _RESOLVE, ResolveParams } from './pattern';
