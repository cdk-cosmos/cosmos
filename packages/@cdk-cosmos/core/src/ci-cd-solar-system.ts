import { Construct, StackProps } from '@aws-cdk/core';

import { Project } from '@aws-cdk/aws-codebuild';
import { Role } from '@aws-cdk/aws-iam';
import {
  PATTERN,
  Galaxy,
  EcsSolarSystem,
  CiCdSolarSystem,
  GalaxyExtension,
  EcsSolarSystemStack,
  EcsSolarSystemProps,
  ImportedEcsSolarSystem,
  EcsSolarSystemExtensionStack,
  CiCdSolarSystemExtension,
  CdkPipeline,
  CdkPipelineProps,
} from '.';

export interface CiCdStackProps extends EcsSolarSystemProps {
  cdkPipelineProps?: Partial<CdkPipelineProps>;
}

export class CiCdSolarSystemStack extends EcsSolarSystemStack implements CiCdSolarSystem {
  readonly CdkDeploy: Project;

  constructor(galaxy: Galaxy, props?: CiCdStackProps) {
    super(galaxy, 'CiCd', props);

    const { cdkPipelineProps = {} } = props || {};
    const { CdkRepo, CdkMasterRoleStaticArn } = this.Galaxy.Cosmos;

    const pipeline = new CdkPipeline(this, 'CdkPipeline', {
      name: this.RESOLVE(PATTERN.SINGLETON_COSMOS, 'Cdk-Pipeline'),
      cdkRepo: CdkRepo,
      deployRole: Role.fromRoleArn(this, 'CdkMasterRole', CdkMasterRoleStaticArn, { mutable: false }),
      // deployVpc: this.Vpc,
      deployStacks: [this.RESOLVE(PATTERN.COSMOS, '*')],
      ...cdkPipelineProps,
    });
    this.CdkDeploy = pipeline.Deploy;

    // RemoteBuildProject.export(`Core${this.Account.Name}${this.Name}`, this.CdkDeploy);
  }
}

export class ImportedCiCdSolarSystem extends ImportedEcsSolarSystem implements CiCdSolarSystem {
  // readonly CdkDeploy: IProject;

  constructor(scope: Construct, galaxy: Galaxy) {
    super(scope, galaxy, 'CiCd');

    // this.CdkDeploy = RemoteBuildProject.import(this, `Core${this.Galaxy.Name}${this.Name}`, 'CdkPipelineDeploy');
  }
}

export class CiCdSolarSystemExtensionStack extends EcsSolarSystemExtensionStack implements CiCdSolarSystemExtension {
  readonly Portal: EcsSolarSystem;
  readonly DeployPipeline: CdkPipeline;
  readonly DeployProject: Project;

  constructor(galaxy: GalaxyExtension, props?: StackProps) {
    super(galaxy, 'CiCd', props);

    this.node.tryRemoveChild(this.Portal.node.id);
    this.Portal = new ImportedCiCdSolarSystem(this, this.Galaxy.Portal);

    this.DeployPipeline = new CdkPipeline(this, 'CdkPipeline', {
      name: `App-${this.Galaxy.Cosmos.Name}-Cdk-Pipeline`,
      cdkRepo: this.Galaxy.Cosmos.CdkRepo,
      deployRole: Role.fromRoleArn(this, 'CdkMasterRole', this.Galaxy.Cosmos.Portal.CdkMasterRoleStaticArn, {
        mutable: false,
      }),
      deployStacks: [this.RESOLVE(PATTERN.COSMOS, '*')],
    });
    this.DeployProject = this.DeployPipeline.Deploy;
  }
}
