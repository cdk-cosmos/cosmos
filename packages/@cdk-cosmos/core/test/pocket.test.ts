import '@aws-cdk/assert/jest';
import { App, Stack } from '@aws-cdk/core';
import { synthesizeStacks } from '../../../../src/test';
import { BaseStack } from '../src/base';
import { Bucket } from '@aws-cdk/aws-s3';

const app = new App();

const stack = new BaseStack(app, 'Devops', {
  type: 'Cosmos',
  partition: 'Core',
});

new Bucket(stack, 'Bucket');
new Bucket(stack, 'Bucket2');

const [synth] = synthesizeStacks(stack);

test('should be a stack', () => {
  console.log(synth.stackName);
  expect(synth.template).toMatchSnapshot();
});
