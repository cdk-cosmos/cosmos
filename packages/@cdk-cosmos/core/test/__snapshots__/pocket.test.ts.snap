// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be a stack 1`] = `
Object {
  "Resources": Object {
    "Bucket": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "Bucket2": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
